clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -32,11 +36,29 @@
  * 
  * Metadata is maintained for only a subset of topics, which can be added to over time. When we request metadata for a
  * topic we don't have any metadata for it will trigger a metadata update.
+ * <p>
+ * Expiry policy of topics are set when the topics are added to the metadata.
+ * <ul>
+ *   <li> Topics added without expiry using {@link #setTopicsWithoutExpiry(Collection)} are retained until they are explicitly removed
+ *        using the same method. Consumers use this policy to manage topics without relying on automated expiry.</li>
+ *   <li> Topics added with expiry using {@link #addWithExpiry(String)} are retained for one refresh interval after the last
+ *        add. Producers use this policy to remove topics when no there are no attempts to send messages to the topics
+ *        for one refresh interval.</li>
+ *   <li> Unknown topics are removed from metadata if expiry is enabled, to avoid unnecessary metadata refresh of deleted topics
+ *        when the topic is no longer in use. The topic is added again when required, for example, if there is a message to send.</li>
+ * </ul>
  */
 public final class Metadata {
 
     private static final Logger log = LoggerFactory.getLogger(Metadata.class);
 
+    /* Producer sets topic expiry to this value when a message is available to send. Expiry is updated on next metadata refresh */
+    private static final Long TOPIC_EXPIRY_NEEDS_UPDATE = -1L;
+    /* Metadata request returned UNKNOWN_TOPIC_OR_PARTITION, topic is expired on next metadata refresh if not used within the interval */
+    private static final Long TOPIC_EXPIRY_PENDING = -2L;
+    /* Consumers and consumer coordinator do not need automated expiry since topics are managed independently by the coordinator */
```

Comment:
I think some of the complexity in this patch comes from the fact that we're trying to support the producer and consumer use cases transparently within the same instance. I wonder if it would be easier to allow a flag to be passed to the constructor to enable or disable cache expiration?